Add convenience `::expectFilterRemoved()` `::expectActionRemoved()` Fix #157

BrianHenryIE opened the pull request:
# Summary

As said in #157:

> WP_Mock provides useful helpers for `expectFilterAdded` and `expectFilterNotAdded`, and [their] companion actions. However, we have no canonical way of checking `remove_action` nor `remove_filter` calls.

See: [`remove_filter()`](https://github.com/WordPress/WordPress/blob/5b2ef0f91d4ee08703e755a42f3c2e96e36bdf0a/wp-includes/plugin.php#L313).

### Closes: #157

## Details

It's really just a convenience function on top of `WP_Mock::userFunction()`.

```php
public static function expectFilterRemoved(string $filter, $callback, $priority = null) : void
{
    self::userFunction(
        'remove_filter'
        array(
            'args'   => array_filter(func_get_args()),
            'times'  => 1,
        )
    );
}
```

We're using `array_filter()` there to handle the optional use of `$priority` in `remove_filter()`.

`::expectFilterNotRemoved()` is the same with `'times' => 0`.

## Contributor checklist 

- [x] I agree to follow this project's [**Code of Conduct**](https://github.com/10up/.github/blob/trunk/CODE_OF_CONDUCT.md).
- [x] I have updated the documentation accordingly 
- [x] I have added tests to cover changes introduced by this pull request
- [x] All new and existing tests pass

## Testing


### Reviewer checklist <!-- Required -->

<!-- The following checklist is for the reviewer: add any steps that may be relevant while reviewing this pull request --> 

- [ ] Code changes review
- [ ] Documentation changes review
- [ ] Unit tests pass
- [ ] Static analysis passes

BrianHenryIE committed:
Add convenience functions for `:: expectActionRemoved()` `::expectFilterRemoved()`

BrianHenryIE committed:
Rename `$action` to `$filter` in filter related function

BrianHenryIE committed:
Add expectHookRemoved tests

BrianHenryIE committed:
Add expectHookNotRemoved functions

BrianHenryIE committed:
Remove strict typing to allow `::type( 'int' )`

`\WP_Mock\Functions::type( 'int' )`

BrianHenryIE committed:
update functions' type annotations

BrianHenryIE committed:
Add doc: Asserting that actions and filters have been removed

nmolham-godaddy committed:
Merge branch 'trunk' into add/remove-action-remove-filter

nmolham-godaddy commented on php/WP_Mock.php:
```diff
@@ -546,4 +546,88 @@ public static function getDeprecatedMethodListener(): DeprecatedMethodListener
+     * @param string|callable-string|callable|Type $callback the callable to be removed
+     * @param int|Type|null $priority the priority it should be registered at
+     *
+     * @return void
+     * @throws InvalidArgumentException
+     */
+    public static function expectActionRemoved(string $action, $callback, $priority = null) : void
```
How about returning the expectation once set? incase it was needed to for later. As well as adding nullable-int data type for the `$priority` argument

```suggestion
     * @return Mockery\Expectation
     * @throws InvalidArgumentException
     */
    public static function expectActionRemoved(string $action, $callback, ?int $priority = null) : Mockery\Expectation
```

nmolham-godaddy commented on php/WP_Mock.php:
```diff
@@ -546,4 +546,88 @@ public static function getDeprecatedMethodListener(): DeprecatedMethodListener
+     */
+    public static function expectActionRemoved(string $action, $callback, $priority = null) : void
+    {
+        self::userFunction(
+            'remove_action',
+            array(
+                'args'   => array_filter(func_get_args()),
+                'times'  => 1,
+            )
+        );
```
For better performance, it would be ideal to be more implicit about the aguments instead of reaching for the `func_get_args`, that way the interpreter won't have to do much assumption work with the data types.

Also, we are trying to move away from the classic array syntax towards.

```suggestion
        $args = [$action, $callback];
        if ($priority) {
            $args[] = $priority;
        }

        return self::userFunction('remove_action', [
            'args'  => $args,
            'times' => 1,
        ]);
```

nmolham-godaddy reviewed (changes requested):
Thanks @BrianHenryIE for this nice PR, I have a couple of suggestions below, which I added for one method, but apply for all of the other new methods as well.

nmolham-godaddy committed:
Merge branch 'trunk' into add/remove-action-remove-filter

BrianHenryIE committed:
Update php/WP_Mock.php

Co-authored-by: Nabeel Molham <72820458+nmolham-godaddy@users.noreply.github.com>

BrianHenryIE committed:
Update php/WP_Mock.php

Co-authored-by: Nabeel Molham <72820458+nmolham-godaddy@users.noreply.github.com>

BrianHenryIE committed:
Correct PhpDoc return to match `::setUpMock()`'s `Mockery\Expectation|Mockery\CompositeExpectation`

BrianHenryIE committed:
Apply PR review suggestions

BrianHenryIE committed:
Add `\WP_Mock\Functions::type( 'int' )` suggestion to `expectFilterNotRemoved()`

BrianHenryIE commented:
* `0` is a valid action priority so I changed the suggestion's `if ($priority) {` to an `is_null()` check rather than a falsey check
* I removed the `?int` type on `$priority` to allow `Type` also
* Removed the explicit `Mockery\Expectation` return type and set the PhpDoc to `ExpectationInterface` after tests failed because they returned `CompositeExpectation` (from `Functions::setUpMock()`) and updated the intermediate PhpDoc
* I updated the other functions to apply the suggestions
* Added a documentation line to suggest where using `\WP_Mock\Functions::type( 'int' )` might be helpful

BrianHenryIE committed:
Merge remote-tracking branch 'upstream' into pr/246

BrianHenryIE committed:
PHPStan

BrianHenryIE committed:
Undo unrelated type changes
